### ISSUE TITLE
Idea: draw the last column’s/row’s background on the right/bottom waste area

### ISSUE THREAD

**Issue by chris-morgan** [alacritty/alacritty#3746]
If your window size isn’t a multiple of the block size, you end up with a bit of waste space on the bottom and/or right edges of your window.

Currently Alacritty draws your background colour there, but that might not be the best thing to draw there.

As an extreme example, take this white-on-black terminal currently rendering a black-on-white Vim theme:

![Screenshot showing a conspicuous black band on the right hand edge of the window](https://user-images.githubusercontent.com/392868/82235629-94bbf200-9950-11ea-80b1-747c6ef46238.png)

That black band on the right is ugly. 🙁

What if it took on the background colour of the last block in the column or row, and looked like this instead:

![image](https://user-images.githubusercontent.com/392868/82236539-dd27df80-9951-11ea-900b-26093fb1f370.png)

As a subtlety, you *could* choose to draw the full style of the last character which would get you an underline in the first row, from the X character, but it’s probably harmless either way. Though maybe invert would actually be significant. I dunno, I haven’t thought this through completely.

This could be a mildly risky thing to do, but I think it would achieve a good result almost all of the time.

(I thought of this when planning out hypothetical video streaming, deciding to size my windows exactly and capture client area only, for optimally crisp effect, and then realising that doing it that way I’d get the black band—so I tried filling in the right edge in OBS Studio manually, which kinda works so long as it’s *completely* consistent. On reflection, this sort of issue has bothered me in various terminal emulators for *many* years.)

**Comment by kchibisov** [comment]
I kinda like that idea to draw background of the nearest cell. The only issue I see is that it could be slightly a bit more work for the render, but maybe it's something not that costly? Anyway, it's +1 from me.

**Comment by kchibisov** [comment]
I've updated title to be a bit more clear.

Also, it seem like #2803 is a direct request for coloring, however you're asking for slightly different thing.

Also @zsugabubus  , since you're mentioned that here https://github.com/alacritty/alacritty/issues/2803#issuecomment-546588004 maybe we can upstream your patches if they perform well and simple enough?

**Comment by chrisduerr** [comment]
Why?

I don't see why everyone should put in the extra work to compute the proper color of the extra padding, when this is really just a problem with the themes used by the application running inside of Alacritty.

If your vim colorscheme is completely white and you have a black background configured, without changing the background of Alacritty using the correct escape sequences, you're simply just doing things incorrectly. If you're using the appropriate terminal functionality to try and achieve what you want, this problem does not exist.

**Comment by kchibisov** [comment]
>If your vim colorscheme is completely white and you have a black background configured, without changing the background of Alacritty using the correct escape sequences, you're simply just doing things incorrectly. If you're using the appropriate terminal functionality to try and achieve what you want, this problem does not exist.

It's not only about vim colorscheme tbh, it's also about some status lines on bottom/top, etc. Like there's no way to don't have it flying in the air, and it's not about proper scheme here. In Vim I agree, just use white background for everything, but you can't really force right ways.

![picture](https://user-images.githubusercontent.com/27620401/82252658-1400f400-9958-11ea-9f4f-d1edc944a821.png)

**Comment by chrisduerr** [comment]
> Like there's no way to don't have it flying in the air

I think especially with dynamic padding it looks much better to have these statuslines highlighted as a single line, rather than having the padding below them changed too, making it look like the font alignment is off.

**Comment by kchibisov** [comment]
I mean if you apply dynamic padding + color padding based off content around you it'll look even better.

**Comment by chrisduerr** [comment]
> I mean if you apply dynamic padding + color padding based off content around you it'll look even better.

It won't though. You'll have a line that looks like it should just be a single line with a certain background, but what you'll have is like a line with 1.6 times the normal line height and content that isn't centered but instead aligned to the top. That just doesn't work well.

**Comment by kchibisov** [comment]
>That just doesn't work well.

That's a bit subjective, but I get what you mean anyway. I guess continuation of vim lines is the main benefit for that. Though, I understand folks using light theme with dark terminal.

**Comment by chrisduerr** [comment]
Is there any prior art of this?

**Comment by kchibisov** [comment]
I mean the goal could be to make those lines finished and reach the border correctly. I personally don't like dynamic padding at all.

![picture](https://user-images.githubusercontent.com/27620401/82272841-6952fa80-9984-11ea-918c-f1f404c115ed.png)

**Comment by zsugabubus** [comment]
You cannot always come up with a single padding color that looks good in all cases, for example, colorful `$RPROMPT`s in shells, multicolor statuslines.

---

I updated [my comment](https://github.com/alacritty/alacritty/issues/2803#issuecomment-546588004), so instead of a dead link it points to my hacky patch. I came up with the simplest solution (for me): when a cell is rendered on the edge, it renders one (or two) more for the padding. If you say it worth a PR, I can clean up the code.

I haven’t measured it’s performance penalty.

**Comment by chrisduerr** [comment]
> You cannot always come up with a single padding color that looks good in all cases

If you setup your colorschemes properly, you absolutely can.

> I mean the goal could be to make those lines finished and reach the border correctly as it was stated in the original post.

That has the same issue as extending it vertically though. At that point you've got extra empty space at the side that doesn't look right.

> I personally don't like dynamic padding at all.

Seems like if you don't like the existing and working solutions for that, that's kinda on you? I personally find it to be working very well.

There's also things we cannot use the nearest cell for, no matter what. Like box drawing characters for example. No matter what you do you won't continue those until the end, there'll always be the additional 'padding', regardless of what color you make it. That's the big advantage of the dynamic padding, because it will actually be a consistent padding around the window border, instead of pretending there is no padding when there very clearly is some.

**Comment by zsugabubus** [comment]
> If you setup your colorschemes properly, you absolutely can.

If you have at least two different colors on the right side, gaps must appear somewhere. In the case of Vim, it’s maximum the statusline plus tabline (with one window), but for example my shell’s `$RPROMPT` it means a bunch of lines. 

Things certainly looked prettier with my patch, but no performance penalty worth that much for me, so I do not have a strong opinion about it.

**Comment by kchibisov** [comment]
>There's also things we cannot use the nearest cell for, no matter what. Like box drawing characters for example. No matter what you do you won't continue those until the end, there'll always be the additional 'padding', regardless of what color you make it. That's the big advantage of the dynamic padding, because it will actually be a consistent padding around the window border, instead of pretending there is no padding when there very clearly is some.

That's a good point and I guess dynamic_padding is likely better here. It doesn't solve the issue about multi color schemed setups. Actually I guess vim should set background color to the background it's using inside and reset on exit, etc, so the users won't see that 'ugly' black padding at all. imho. I guess it's something achievable with scripting right?

**Comment by chrisduerr** [comment]
> It doesn't solve the issue about multi color schemed setups.

I'd honestly say that if your colorscheme is using multiple colors, you're probably doing something wrong. I just do not see why you would do that without changing the terminal background, just makes no sense. Even if we would pretend for a second that it's not completely backwards to have multiple different colorschemes in a terminal for different applications.

**Comment by kchibisov** [comment]
>Just to be 100% clear, echo -e "\e]11;rgb:ff/00/ff\e\\" will also color the padding. Everything that modifies the actual terminal background, also modifies the padding color.

Yeah, I've mentioned that, I'm not sure if vim can do that automatically.

**Comment by HiPhish** [comment]
I have been thinking about this issue lately, so I decided to make a mockup.

![mockup](https://github.com/user-attachments/assets/f3269cbb-5bf4-4267-b8bd-0b31c7f9a0aa)

Open the image in a new window or tab to see the details. In my case setting the background colour of the terminal to match my theme is not enough because I have several colours on the edges. My mock-up uses the background colour of the adjacent outer cell as the background of the terminal.

Overall I like the result and I don't mind that the cells are slightly larger on the border, but the there is an issue with the vertical box-drawing character between the time and date: because the box-drawing character is part of the foreground it is not applied to the background of the terminal. I guess you *could* have special treatment for cells with a box-drawing character in them, but what about other fancy characters users come up with in the private use area? Think of the fancy triangles or semi-circles people like to put in their status bar. There is no way of stretching those to fit. So we can either live with a border or cut-off special characters.

**Comment by gustavcedersjo** [comment]
I like the idea, so I made a quick solution to try it out. I think it works well for the space between the last cell to the window border, especially with dynamic padding enabled. But you can get weird effects if you have additional padding.
